# Apriori

### O algoritmo Apriori é um dos métodos mais conhecidos e utilizados para a mineração de padrões frequentes e a geração de regras de associação em bases de dados transacionais. Ele foi introduzido por Rakesh Agrawal e Ramakrishnan Srikant em 1994 e tornou-se a base para muitos outros algoritmos de mineração de dados.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Como Funciona o Algoritmo Apriori
#### O Apriori opera em duas etapas principais: a geração de conjuntos frequentes (itemsets) e a geração de regras de associação. O princípio básico do algoritmo é que qualquer subconjunto de um itemset frequente também deve ser frequente. Isso é conhecido como propriedade Apriori.

## Etapa 1: Geração de Conjuntos Frequentes
#### **-Contagem dos Itens Individuais:** Inicialmente, o algoritmo conta a frequência de cada item individual na base de dados. Mantém apenas os itens que atendem a um suporte mínimo especificado.
#### **-Geração de Conjuntos Candidatos:** Gera conjuntos de itens candidatos de tamanho 2 a partir dos itens frequentes de tamanho 1. Conta a frequência de cada conjunto candidato na base de dados.
#### **-Filtragem de Conjuntos Candidatos:** Mantém apenas os conjuntos candidatos que atendem ao suporte mínimo. Repete o processo para gerar conjuntos candidatos de tamanho 3 a partir dos conjuntos frequentes de tamanho 2, e assim por diante.
#### **-Repetição:** Continua gerando e filtrando conjuntos candidatos até que não restem mais conjuntos candidatos frequentes.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Etapa 2: Geração de Regras de Associação
#### **Formação de Regras:** Para cada conjunto frequente, o algoritmo gera todas as regras possíveis que podem ser derivadas do conjunto. Uma regra é na forma de $A→B$, onde $A$ e $𝐵$ são subconjuntos do itemset frequente e $A∩B=∅$.


#### **-Cálculo de Métricas:** Calcula a confiança de cada regra. A confiança é a proporção de transações que contêm $𝐴$ que também contêm $𝐵$. Se a confiança de uma regra for maior ou igual a um limiar mínimo de confiança especificado, a regra é considerada forte.
### **-Filtragem de Regras:** Mantém apenas as regras que atendem ao limiar de confiança especificado.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

##

## Para aplicar regras de associação utilizando o algoritmo Apriori no dataset do SPAECE 2018, podemos seguir alguns passos. Aqui está um esboço de como isso pode ser feito:



## Passos para Aplicar Apriori no Dataset SPAECE

#### 1- **Preparação dos Dados:**

##### 1.1 - Transformação dos Dados: Como o algoritmo Apriori trabalha melhor com dados categóricos, é importante transformar as variáveis numéricas e categóricas em um formato adequado. Por exemplo, se tivermos colunas como "SEXO", "IDADE", "FORMAÇÃO DA MÃE", etc., podemos transformá-las em valores categóricos binários (one-hot encoding).

##### 1.2 - Discretização: Para variáveis contínuas, podemos discretizá-las em intervalos ou categorias. Por exemplo, as colunas de "NUMERO DE PONTOS EM PORTUGUES" e "NUMERO DE PONTOS EM MATEMATICA" podem ser discretizadas em faixas (por exemplo, baixa, média, alta).

#### 2 - **Transformação em Transações:** Cada linha do dataset representará uma transação e cada coluna representará um item. Precisamos transformar o dataset de modo que cada transação contenha apenas os itens presentes nessa transação específica.

#### 3 - **Aplicação do Algoritmo Apriori:** Usaremos uma biblioteca como 'mlxtend' em Python para aplicar o algoritmo Apriori e extrair regras de associação.

#### 4 - **Interpretação das Regras:** Analisaremos as regras geradas para encontrar padrões interessantes. Por exemplo, podemos descobrir que estudantes com certas características têm maior probabilidade de alcançar altas pontuações em matemática ou português.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [2]:
## Para recuperar o dataset original
gnetDf = pd.read_csv("datasets/SPAECE_2018_codificado.csv", encoding='latin1')

In [3]:
gnetDf

SEXO  IDADE  FORMACAO DA MAE  FORMACAO DO PAI  ACESSO A INTERNET  \
0         1     17                1                3                  0   
1         1     18                0                3                  0   
2         1     17                0                3                  0   
3         1     17                0                3                  0   
4         2     17                0                3                  0   
...     ...    ...              ...              ...                ...   
52489     1     17                3                1                  1   
52490     1     19                3                1                  1   
52491     2     19                3                1                  1   
52492     2     18                3                1                  1   
52493     2     17                3                1                  1   

       NUMERO DE COMPUTADORES  INICIOU OS ESTUDOS  CURSO O ENSINO FUNDAMENTAL  \
0                           1                   3                           1   
1                           1                   5                           1   
2                           1                   5                           1   
3                           1                   3                           3   
4                           1                   5                           1   
...                       ...                 ...                         ...   
52489                       2                   7                           2   
52490                       2                   5                           3   
52491                       2                   5                           1   
52492                       2                   5                           1   
52493                       2                   3                           1   

       REPROVOU  ABANDONOU A ESCOLA  \
0             0                   0   
1             0                   0   
2             0                   0   
3             0                   0   
4             0                   0   
...         ...                 ...   
52489         0                   1   
52490         0                   0   
52491         0                   1   
52492         0                   0   
52493         0                   0   

       GESTAO ESCOLAR / PROFESSORES TRABALHAM COM FOCO NO SPAECE  \
0                                                      2           
1                                                      2           
2                                                      3           
3                                                      1           
4                                                      2           
...                                                  ...           
52489                                                  0           
52490                                                  3           
52491                                                  3           
52492                                                  2           
52493                                                  3           

       PROFESSORES TIRAM DUVIDAS  PROFESSORES CORRIGEM AS ATIVIDADES  \
0                              2                                   2   
1                              3                                   3   
2                              2                                   2   
3                              1                                   1   
4                              2                                   2   
...                          ...                                 ...   
52489                          1                                   1   
52490                          3                                   3   
52491                          2                                   1   
52492                          2                                   2   
52493                          3                       

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [4]:
# Discretização (simplificado)
gnetDf['SEXO_CATEG'] = pd.cut(gnetDf['SEXO'], bins=2, labels=['1', '2'])
gnetDf['IDADE_CATEG'] = pd.cut(gnetDf['IDADE'], bins=4, labels=['17', '18', '19', '20'])
gnetDf['FORMACAO_MAE_CATEG'] = pd.cut(gnetDf['FORMACAO DA MAE'], bins=6, labels=['0','1', '2', '3', '4', '5'])
gnetDf['FORMACAO_PAI_CATEG'] = pd.cut(gnetDf['FORMACAO DO PAI'], bins=6, labels=['0','1', '2', '3', '4', '5'])
gnetDf['ACESSO_INTERNET_CATEG'] = pd.cut(gnetDf['ACESSO A INTERNET'], bins=2, labels=['0','1'])
gnetDf['NUMERO_COMPUTADORES_CATEG'] = pd.cut(gnetDf['NUMERO DE COMPUTADORES'], bins=3, labels=['0','1','2'])
gnetDf['INICIOU_ESTUDOS_CATEG'] = pd.cut(gnetDf['INICIOU OS ESTUDOS'], bins=4, labels=['3','5','7','9'])
gnetDf['ENSINO_FUNDAMENTAL_CATEG'] = pd.cut(gnetDf['CURSO O ENSINO FUNDAMENTAL'], bins=3, labels=['1','2','3'])
gnetDf['REPROVOU_CATEG'] = pd.cut(gnetDf['REPROVOU'], bins=3, labels=['0','1','2'])
gnetDf['ABANDONOU_CATEG'] = pd.cut(gnetDf['ABANDONOU A ESCOLA'], bins=3, labels=['0','1','2'])
gnetDf['GESTAO_CATEG'] = pd.cut(gnetDf['GESTAO ESCOLAR / PROFESSORES TRABALHAM COM FOCO NO SPAECE'], bins=4, labels=['0','1','2','3'])
gnetDf['DUVIDAS_CATEG'] = pd.cut(gnetDf['PROFESSORES TIRAM DUVIDAS'], bins=4, labels=['0','1','2','3'])
gnetDf['CORRIGEM_CATEG'] = pd.cut(gnetDf['PROFESSORES CORRIGEM AS ATIVIDADES'], bins=4, labels=['0','1','2','3'])
gnetDf['TURNO_ESCOLAR_CATEG'] = pd.cut(gnetDf['TURNO ESCOLAR'], bins=4, labels=['1', '2', '3', '4'])

# Conversão dos valores numéricos
gnetDf['PONTOS_PORTUGUES_CATEG'] = pd.cut(gnetDf['NUMERO DE PONTOS EM PORTUGUES'], bins=5, labels=['1', '2', '3', '4', '5'])
gnetDf['PONTOS_MATEMATICA_CATEG'] = pd.cut(gnetDf['NUMERO DE PONTOS EM MATEMATICA'], bins=5, labels=['1', '2', '3', '4', '5'])
gnetDf['PROF_PORTUGUES_CATEG'] = pd.cut(gnetDf['PROFICIENCIA EM PORTUGUES'], bins=5, labels=['1', '2', '3', '4', '5'])
gnetDf['PROF_MATEMATICA_CATEG'] = pd.cut(gnetDf['PROFICIENCIA EM MATEMATICA'], bins=5, labels=['1', '2', '3', '4', '5'])


#### Mapear os valores em conjuntos categóricos para aplicar a divisão binária e também tornar o dataset aplicável para problemas de classificação.

In [5]:
gnetDf

SEXO  IDADE  FORMACAO DA MAE  FORMACAO DO PAI  ACESSO A INTERNET  \
0         1     17                1                3                  0   
1         1     18                0                3                  0   
2         1     17                0                3                  0   
3         1     17                0                3                  0   
4         2     17                0                3                  0   
...     ...    ...              ...              ...                ...   
52489     1     17                3                1                  1   
52490     1     19                3                1                  1   
52491     2     19                3                1                  1   
52492     2     18                3                1                  1   
52493     2     17                3                1                  1   

       NUMERO DE COMPUTADORES  INICIOU OS ESTUDOS  CURSO O ENSINO FUNDAMENTAL  \
0                           1                   3                           1   
1                           1                   5                           1   
2                           1                   5                           1   
3                           1                   3                           3   
4                           1                   5                           1   
...                       ...                 ...                         ...   
52489                       2                   7                           2   
52490                       2                   5                           3   
52491                       2                   5                           1   
52492                       2                   5                           1   
52493                       2                   3                           1   

       REPROVOU  ABANDONOU A ESCOLA  ...  REPROVOU_CATEG  ABANDONOU_CATEG  \
0             0                   0  ...               0                0   
1             0                   0  ...               0                0   
2             0                   0  ...               0                0   
3             0                   0  ...               0                0   
4             0                   0  ...               0                0   
...         ...                 ...  ...             ...              ...   
52489         0                   1  ...               0                1   
52490         0                   0  ...               0                0   
52491         0                   1  ...               0                1   
52492         0                   0  ...               0                0   
52493         0                   0  ...               0                0   

       GESTAO_CATEG  DUVIDAS_CATEG  CORRIGEM_CATEG  TURNO_ESCOLAR_CATEG  \
0                 2              2               2                    2   
1                 2              3               3                    2   
2                 3              2               2                    2   
3                 1              1               1                    2   
4                 2              2               2                    2   
...             ...            ...             ...                  ...   
52489             0              1               1                    4   
52490             3              3               3                    4   
52491             3              2               1                    4   
52492             2              2               2                    4   
52493             3              3               3                    4   

       PONTOS_PORTUGUES_CATEG  PONTOS_MATEMATICA_CATEG PROF_PORTUGUES_CATEG  \
0                           3                        2                    3   
1                           4                        2                    3   
2                           2                        2                

##### Mapeia variáveis categóricas para valores numéricos binários (0 ou 1) utilizando 'pd.get_dummies' para fazer one-hot encoding, transformando as variáveis categóricas em colunas binárias.

In [6]:
#Remover as colunas originais e deixar apenas as categóricas
colunas_a_remover = ['SEXO', 'IDADE', 'FORMACAO DA MAE', 'FORMACAO DO PAI',
       'ACESSO A INTERNET', 'NUMERO DE COMPUTADORES', 'INICIOU OS ESTUDOS',
       'CURSO O ENSINO FUNDAMENTAL', 'REPROVOU', 'ABANDONOU A ESCOLA',
       'GESTAO ESCOLAR / PROFESSORES TRABALHAM COM FOCO NO SPAECE',
       'PROFESSORES TIRAM DUVIDAS', 'PROFESSORES CORRIGEM AS ATIVIDADES',
       'TURNO ESCOLAR', 'NUMERO DE PONTOS EM PORTUGUES',
       'NUMERO DE PONTOS EM MATEMATICA', 'PROFICIENCIA EM PORTUGUES',
       'PROFICIENCIA EM MATEMATICA']
gnetDf = gnetDf.drop(columns=colunas_a_remover)

In [7]:
# Salvar o DataFrame codificado
gnetDf.to_csv("datasets/CLASS.csv", index=False)

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [8]:
df_encoded = pd.get_dummies(gnetDf)

In [9]:
df_encoded

SEXO_CATEG_1  SEXO_CATEG_2  IDADE_CATEG_17  IDADE_CATEG_18  \
0              True         False            True           False   
1              True         False           False            True   
2              True         False            True           False   
3              True         False            True           False   
4             False          True            True           False   
...             ...           ...             ...             ...   
52489          True         False            True           False   
52490          True         False           False           False   
52491         False          True           False           False   
52492         False          True           False            True   
52493         False          True            True           False   

       IDADE_CATEG_19  IDADE_CATEG_20  FORMACAO_MAE_CATEG_0  \
0               False           False                 False   
1               False           False                  True   
2               False           False                  True   
3               False           False                  True   
4               False           False                  True   
...               ...             ...                   ...   
52489           False           False                 False   
52490            True           False                 False   
52491            True           False                 False   
52492           False           False                 False   
52493           False           False                 False   

       FORMACAO_MAE_CATEG_1  FORMACAO_MAE_CATEG_2  FORMACAO_MAE_CATEG_3  ...  \
0                      True                 False                 False  ...   
1                     False                 False                 False  ...   
2                     False                 False                 False  ...   
3                     False                 False                 False  ...   
4                     False                 False                 False  ...   
...                     ...                   ...                   ...  ...   
52489                 False                 False                  True  ...   
52490                 False                 False                  True  ...   
52491                 False                 False                  True  ...   
52492                 False                 False                  True  ...   
52493                 False                 False                  True  ...   

       PROF_PORTUGUES_CATEG_1  PROF_PORTUGUES_CATEG_2  PROF_PORTUGUES_CATEG_3  \
0                       False                   False                    True   
1                       False                   False                    True   
2                       False                    True                   False   
3                       False                   False                    True   
4                       False                   False                   False   
...                       ...                     ...                     ...   
52489                   False                   False                    True   
52490                   False                   False                   False   
52491                   False                   False                   False   
52492                   False                   False                   False   
52493                   False                   False                   False   

       PROF_PORTUGUES_CATEG_4  PROF_PORTUGUES_CATEG_5  \
0                       False                   False   
1                       False                   False   
2                       False                   False   
3                       False                   False   
4                        True                   False   
...                       ...                     ...   
52489                   False                   False   
52490                    True  

In [10]:
df_encoded

SEXO_CATEG_1  SEXO_CATEG_2  IDADE_CATEG_17  IDADE_CATEG_18  \
0              True         False            True           False   
1              True         False           False            True   
2              True         False            True           False   
3              True         False            True           False   
4             False          True            True           False   
...             ...           ...             ...             ...   
52489          True         False            True           False   
52490          True         False           False           False   
52491         False          True           False           False   
52492         False          True           False            True   
52493         False          True            True           False   

       IDADE_CATEG_19  IDADE_CATEG_20  FORMACAO_MAE_CATEG_0  \
0               False           False                 False   
1               False           False                  True   
2               False           False                  True   
3               False           False                  True   
4               False           False                  True   
...               ...             ...                   ...   
52489           False           False                 False   
52490            True           False                 False   
52491            True           False                 False   
52492           False           False                 False   
52493           False           False                 False   

       FORMACAO_MAE_CATEG_1  FORMACAO_MAE_CATEG_2  FORMACAO_MAE_CATEG_3  ...  \
0                      True                 False                 False  ...   
1                     False                 False                 False  ...   
2                     False                 False                 False  ...   
3                     False                 False                 False  ...   
4                     False                 False                 False  ...   
...                     ...                   ...                   ...  ...   
52489                 False                 False                  True  ...   
52490                 False                 False                  True  ...   
52491                 False                 False                  True  ...   
52492                 False                 False                  True  ...   
52493                 False                 False                  True  ...   

       PROF_PORTUGUES_CATEG_1  PROF_PORTUGUES_CATEG_2  PROF_PORTUGUES_CATEG_3  \
0                       False                   False                    True   
1                       False                   False                    True   
2                       False                    True                   False   
3                       False                   False                    True   
4                       False                   False                   False   
...                       ...                     ...                     ...   
52489                   False                   False                    True   
52490                   False                   False                   False   
52491                   False                   False                   False   
52492                   False                   False                   False   
52493                   False                   False                   False   

       PROF_PORTUGUES_CATEG_4  PROF_PORTUGUES_CATEG_5  \
0                       False                   False   
1                       False                   False   
2                       False                   False   
3                       False                   False   
4                        True                   False   
...                       ...                     ...   
52489                   False                   False   
52490                    True  

In [11]:
# Salvar o DataFrame codificado
df_encoded.to_csv("datasets/Apriori.csv", index=False)

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

##### A função apriori da biblioteca 'mlxtend' encontra conjuntos frequentes e gera regras de associação com a função association_rules.

In [12]:
# Aplicar Apriori para encontrar conjuntos frequentes
frequent_itemsets = apriori(df_encoded, min_support=0.1, use_colnames=True)

In [13]:
# Gerar regras de associação
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

In [14]:
# Exibir regras geradas
print(rules)

                                             antecedents  \
0                                         (SEXO_CATEG_1)   
1                                         (SEXO_CATEG_1)   
2                                         (SEXO_CATEG_1)   
3                                         (SEXO_CATEG_1)   
4                                         (SEXO_CATEG_2)   
...                                                  ...   
11092  (INICIOU_ESTUDOS_CATEG_3, DUVIDAS_CATEG_2, GES...   
11093  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11094  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11095  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11096  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, REPR...   

                                           consequents  antecedent support  \
0                            (ACESSO_INTERNET_CATEG_1)            0.532804   
1                         (ENSINO_FUNDAMENTAL_CATEG_1)            0.532804   
2                                   (REPROVOU

### A saída das __*rules*__ do algoritmo Apriori mostra as regras de associação encontradas no dataset, cada uma composta por um conjunto de colunas (atributos) que frequentemente aparecem juntos. A tabela contém as seguintes colunas:

#### **- antecedents:** Conjunto de itens que, quando presentes, aumentam a probabilidade dos itens do consequents estarem presentes.
#### **- consequents:** Conjunto de itens cuja presença é aumentada pela presença dos antecedents.
#### **- support:** Frequência com que a combinação de antecedents e consequents aparece no dataset, medida como a proporção do total de transações.
#### **- confidence:** Probabilidade de encontrar os consequents nas transações que contêm os antecedents. Calculado como 

$$ 
confidence(X→Y)= 
\frac{support(X)}
{support(X∪Y)}
$$

#### **- lift:** Medida de quão frequentemente os consequents ocorrem nas transações que contêm os antecedents em comparação com a frequência esperada, assumindo que os antecedents e consequents são independentes. Calculado como: 

$$
lift(X→Y)= \frac{
support(Y)}
{confidence(X→Y)}
$$​


-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [15]:
# Analisar algumas regras interessantes
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

                                             antecedents  \
0                                         (SEXO_CATEG_1)   
1                                         (SEXO_CATEG_1)   
2                                         (SEXO_CATEG_1)   
3                                         (SEXO_CATEG_1)   
4                                         (SEXO_CATEG_2)   
...                                                  ...   
11092  (INICIOU_ESTUDOS_CATEG_3, DUVIDAS_CATEG_2, GES...   
11093  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11094  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11095  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, CORR...   
11096  (INICIOU_ESTUDOS_CATEG_3, GESTAO_CATEG_2, REPR...   

                                           consequents   support  confidence  \
0                            (ACESSO_INTERNET_CATEG_1)  0.385492    0.723515   
1                         (ENSINO_FUNDAMENTAL_CATEG_1)  0.436126    0.818549   
2                                   (RE

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Análise das Regras de Associação

### Exemplo 1: Primeira Regra
#### **-antecedents:** (SEXO_CATEG_1)
#### **-consequents:** (ACESSO_INTERNET_CATEG_1)
#### **-support:** 0.385492
#### **-confidence:** 0.723515
#### **-lift:** 0.990099

### Interpretação:

#### **-antecedents:** A presença da categoria 1 da coluna SEXO (que mapeamos como Feminino).
#### **-consequents:** Está associada à presença da categoria 1 da coluna ACESSO A INTERNET (que mapeamos como Sim).
#### **-support:** Cerca de 38.55% das transações (linhas) no dataset contêm tanto 'Feminino' como 'Sim' para acesso à internet.
#### **-confidence:** Em 72.35% das vezes que a coluna SEXO é 'Feminino', a coluna ACESSO A INTERNET é 'Sim'.
#### **-lift:** O valor de 0.99 indica que há quase nenhuma relação entre SEXO ser 'Feminino' e ACESSO A INTERNET ser 'Sim', pois está muito próximo de 1 (independência).

##

##

### Exemplo 2: Segunda Regra
#### **-antecedents:** (SEXO_CATEG_1)
#### **-consequents:** (ENSINO_FUNDAMENTAL_CATEG_1)
#### **-support:** 0.436126
#### **-confidence:** 0.818549
#### **-lift:** 1.001653

### Interpretação:

#### **-antecedents:** A presença da categoria 1 da coluna SEXO (Feminino).
#### **-consequents:** Está associada à presença da categoria 1 da coluna ENSINO FUNDAMENTAL (Ensino Fundamental Completo).
#### **-support:** Cerca de 43.61% das transações contêm tanto 'Feminino' como 'Ensino Fundamental Completo'.
#### **-confidence:** Em 81.85% das vezes que a coluna SEXO é 'Feminino', a coluna ENSINO FUNDAMENTAL é 'Completo'.
#### **-lift:** O valor de 1.001653 sugere uma relação quase neutra, muito próximo de independência.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Observar

### **- Transformação dos Dados:** Cada valor no DataFrame final deve ser binário para que o algoritmo Apriori funcione corretamente.
### **- Interpretação das Regras:** As regras geradas podem ser analisadas para descobrir padrões interessantes nas respostas dos alunos.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Aplicar

### **- Identificação de Perfis de Estudantes:** Descobrir perfis de estudantes que têm maior probabilidade de alcançar altas pontuações, permitindo intervenções direcionadas.
### **- Apoio à Decisão Educacional:** Fornecer dados para políticas educacionais e práticas de ensino baseadas em padrões de desempenho dos estudantes.
### **- Análise de Comportamentos de Estudo:** Entender como diferentes comportamentos e características dos estudantes influenciam seu desempenho acadêmico.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Vantagens do Apriori
### **-Simplicidade:** Fácil de entender e implementar.
### **-Popularidade:** Amplamente utilizado e bem estudado na literatura de mineração de dados.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Desvantagens do Apriori
### **-Ineficiente para Grandes Conjuntos de Dados:** Gera muitos candidatos e requer múltiplas varreduras na base de dados.
### **-Custo Computacional:** Pode ser computacionalmente caro devido à geração e verificação de muitos candidatos.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------